webgpu.h

```diff
@@ -144,6 +144,22 @@ typedef enum WGPUCullMode {
     WGPUCullMode_Force32 = 0x7FFFFFFF
 } WGPUCullMode;
 
+typedef enum WGPUErrorFilter {
+    WGPUErrorFilter_None = 0x00000000,
+    WGPUErrorFilter_Validation = 0x00000001,
+    WGPUErrorFilter_OutOfMemory = 0x00000002,
+    WGPUErrorFilter_Force32 = 0x7FFFFFFF
+} WGPUErrorFilter;
+
+typedef enum WGPUErrorType {
+    WGPUErrorType_None = 0x00000000,
+    WGPUErrorType_Validation = 0x00000001,
+    WGPUErrorType_OutOfMemory = 0x00000002,
+    WGPUErrorType_Unknown = 0x00000003,
+    WGPUErrorType_ContextLost = 0x00000004,
```

Comment:
I'm not sure this is actually useful. It's only needed because the DeviceLost callback currently receives one. But I don't think we would ever call it with any other value (even WGPUErrorType_None) - we should always call this callback even if the device is being intentionally destroyed. Basically the DeviceLost callback would be a "device destroyed for any reason" callback.

On the Web, we might not want to resolve `device.lost` if the page is in the middle of shutdown. But I think that logic should be at the browser level.

Reply:
So just to be sure that enum value would disappear in favor of a specific callback type for device destroyed that's like a Rust `FnOnce() -> ()`? SGTM. Alternatively or in addition we could have a getter for device loss.

This also means that for testing we'll need the equivalent of WebGL's "lose context" extension.

Reply:
Mmmmh good point but that doesn't match the [ErrorHandling](https://github.com/gpuweb/gpuweb/blob/master/design/ErrorHandling.md) explainer. Maybe that can be a small divergence between this header and JS. Translation would do something like this:

 - JS API on C API: ignore device lost (or validate the lost handler is called before any device lost error).
 - C API on JS API: every popErrorScope checks whether the device is lost before calling the callback.

Reply:
Ah, rejection is like the context lost so it makes sense to keep the enum.

However maybe rename context lost to device lost because we aren't in WebGL? ^^

Reply:
We also have
> If there are no error scopes on the stack, `popErrorScope()` rejects.

If we just call the callback with `ERROR_TYPE_VALIDATION`, then for JS on C, then an error due to an empty stack looks the same as a validation error contained within a scope.

Maybe we should have `popErrorScope` return a success boolean?

Reply:
hm, for memory management it is still probably more convenient for us to call that callback. I guess we could just use UNKNOWN?

In JS we could change it to throw an exception if you pop an empty stack, and the best way to expose that would probably be by returning a bool. If we did that then native apps would have to either do a little extra memory management or just ASSERT and allow it to leak in release.

Reply:
UNKNOWN would work, but the naming is a bit cryptic.